File: shape_detection/shape_detection.py
from optparse import OptionParser
import cv2
import os
from numpy import poly

# Apply contour over shape, add color and text
def draw_shape(image, contour, shape_name, color):
    cv2.drawContours(image,[contour],0,color,-1)
     # Find contour center to place text at the center
    M = cv2.moments(contour)
    cx = int(M['m10'] / M['m00'])
    cy = int(M['m01'] / M['m00'])
    cv2.putText(image, shape_name, (cx-50, cy), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 0, 0), 1)

def detect_shapes(input_file, output_file):
    # Read image file
    image = cv2.imread(input_file)
    # Convert image to grayscale
    gray_image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    # Apply threshold to image
    ret, threshold_image = cv2.threshold(gray_image, 127, 255, cv2.THRESH_BINARY)
    # Get contours from image
    _, contours, hierarcy = cv2.findContours(threshold_image, cv2.RETR_LIST, cv2.CHAIN_APPROX_NONE)
    
    # Removing top level contour
    n = len(contours) - 1
    contours = sorted(contours, key=cv2.contourArea, reverse=False)[:n]
    
    for contour in contours:
        # Get closest polygon matching countour
        epsilon = 0.01 * cv2.arcLength(contour, True)
        poly_sides = cv2.approxPolyDP(contour, epsilon, True)
       
        # Count number of sides of polygon and decide shape
        if len(poly_sides) == 3:
            draw_shape(image, contour, "Triangle", (255, 0, 0))
        elif len(poly_sides) == 4:
            x,y,w,h = cv2.boundingRect(contour)
            if abs(w-h) <=3:
                draw_shape(image, contour, "Square", (0,255,0))
            else:
                draw_shape(image, contour, "Rectangle", (0,0,255))
        elif len(poly_sides) == 10:
            draw_shape(image, contour, "Star", (255,255,0))
        elif len(poly_sides) >= 15:
            draw_shape(image, contour, "Circle", (0,255,255))
        
    cv2.imwrite(output_file, image)

if __name__ == '__main__':
    input_file = 'someshapes.jpg'
    output_file = 'labeled_shapes.jpg'
    detect_shapes(input_file, output_file)

File: shape_detection/src/shape_detection.cpp
//============================================================================
// Name        : shape_detection.cpp
// Author      : 
// Version     :
// Copyright   : Your copyright notice
// Description : Hello World in C++, Ansi-style
//============================================================================

#include <iostream>
#include <string>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

void printShape(Mat& image, vector<Point>& contour, string& name, Scalar& color){
	vector<vector<Point>> contours;
	contours.push_back(contour);
	drawContours(image,contours,0,color,-1);
	Moments moment = moments(contour, false);
	int cx = int(moment.m10/moment.m00);
	int cy = int(moment.m01/moment.m00);
	Point center(cx-60, cy);
	Scalar black(0, 0, 0);
	putText(image, name, center, FONT_HERSHEY_COMPLEX, 1, black, 1);
}

void detectShape(const string& inputFile, const string& outputFile){
	// Reading input image file
	Mat input;
	input = imread(inputFile, CV_LOAD_IMAGE_COLOR);

	imshow("Input Image", input);
	waitKey(0);

	// Convert image to gray scale
	Mat grayImage;
	cvtColor(input, grayImage, COLOR_BGR2GRAY);

	// Get threshold value from image
	Mat thresholdImage;
	threshold(grayImage, thresholdImage, 70, 255, THRESH_BINARY);

	// Get contours
	vector<vector<Point> > contours;
	findContours(thresholdImage, contours, RETR_LIST, CHAIN_APPROX_NONE);

	//Remove top level bounding contour
	sort(contours.begin(), contours.end(),
			[](const vector<Point>& c1, const vector<Point>& c2){
				return contourArea(c1, false) < contourArea(c2, false);
			}
	);
	contours.pop_back();

	// Iterate over contours
	for(auto& contour: contours){
		// Detect closest polygon
		vector<Point> polySides;
		double epsilon = 0.01 * arcLength(contour, true);
		approxPolyDP(contour, polySides, epsilon, true);

		// Label image with polygon name
		int sides = polySides.size();
		if(sides == 3){
			cout << "Found a triangle";
			Scalar color(255,0,0);
			string name = "Triangle";
			printShape(input, contour, name, color);
		}else if(sides == 4){
			Rect rect = boundingRect(contour);
			if(abs(rect.width - rect.height) <= 3){
				string name = "Square";
				Scalar color(0,255,0);
				printShape(input, contour, name, color);
			}else{
				string name = "Rectangle";
				Scalar color(0,0,255);
				printShape(input, contour, name, color);
			}
		}else if(sides == 10){
			string name = "Star";
			Scalar color(255,255,0);
			printShape(input, contour, name, color);
		}else if(sides >= 15){
			string name = "Circle";
			Scalar color(0, 255,255);
			printShape(input, contour, name, color);
		}
	}

	// Save output image
	imwrite(outputFile, input);
}

int main(int argc, char** argv) {
	string inputFile = "../someshapes.jpg";
	string outputFile = "../labeled_shapes.jpg";

	detectShape(inputFile, outputFile);

	return 0;
}
